Riot/Modules/Room/VoiceMessages/VoiceMessageController.swift

```diff
@@ -0,0 +1,416 @@
+    
+    private enum Constants {
+        static let maximumAudioRecordingDuration: TimeInterval = 120.0
+        static let maximumAudioRecordingLengthReachedThreshold: TimeInterval = 10.0
+        static let minimumRecordingDuration = 1.0
+    }
+    
+    private let themeService: ThemeService
+    private let mediaServiceProvider: VoiceMessageMediaServiceProvider
+    private let temporaryFileURL: URL
+    
+    private let _voiceMessageToolbarView: VoiceMessageToolbarView
+    private var displayLink: CADisplayLink!
+    
+    private var audioRecorder: VoiceMessageAudioRecorder?
+    
+    private var audioPlayer: VoiceMessageAudioPlayer?
+    private var waveformAnalyser: WaveformAnalyzer?
+    
+    private var audioSamples: [Float] = []
+    private var isInLockedMode: Bool = false
+    private var notifiedRemainingTime = false
+    
+    @objc public weak var delegate: VoiceMessageControllerDelegate?
+    
+    @objc public var isRecordingAudio: Bool {
+        return audioRecorder?.isRecording ?? false || isInLockedMode
+    }
+    
+    @objc public var voiceMessageToolbarView: UIView {
+        return _voiceMessageToolbarView
+    }
+    
+    @objc public init(themeService: ThemeService, mediaServiceProvider: VoiceMessageMediaServiceProvider) {
+        self.themeService = themeService
+        self.mediaServiceProvider = mediaServiceProvider
+        
+        let temporaryDirectoryURL = URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
+        temporaryFileURL = temporaryDirectoryURL.appendingPathComponent(ProcessInfo().globallyUniqueString).appendingPathExtension("m4a")
+        
+        _voiceMessageToolbarView = VoiceMessageToolbarView.loadFromNib()
+        
+        super.init()
+        
+        _voiceMessageToolbarView.delegate = self
+        
+        displayLink = CADisplayLink(target: WeakTarget(self, selector: #selector(handleDisplayLinkTick)), selector: WeakTarget.triggerSelector)
+        displayLink.isPaused = true
+        displayLink.add(to: .current, forMode: .common)
+        
+        NotificationCenter.default.addObserver(self, selector: #selector(updateTheme), name: .themeServiceDidChangeTheme, object: nil)
+        updateTheme()
+        
+        updateUI()
+    }
+    
+    // MARK: - VoiceMessageToolbarViewDelegate
+    
+    func voiceMessageToolbarViewDidRequestRecordingStart(_ toolbarView: VoiceMessageToolbarView) {
+        guard AVAudioSession.sharedInstance().recordPermission == .granted else {
+            delegate?.voiceMessageControllerDidRequestMicrophonePermission(self)
+            return
+        }
+        
+        // Haptic are not played during record on iOS by default. This fix works
+        // only since iOS 13. A workaround for iOS 12 and earlier would be to
+        // dispatch after at least 100ms recordWithOuputURL call
+        if #available(iOS 13.0, *) {
+            try? AVAudioSession.sharedInstance().setCategory(.playAndRecord)
```

Comment:
This is set here and again below in `audioRecorder.recordWithOuputURL`. And I may be wrong here, but I always understood that you have to call `AVAudioSession.sharedInstance().setActive` for this category to have an effect (disabling it after you're done).

Reply:
You're right, good catch 👍

Reply:
We do need to keep this one here though otherwise you won't get haptics on the first invocation.